idea/src/hev-timeline/hev-interpolateScale.1
.TH hev-interpolateScale 1 "November 2011"

.SH NAME
hev-interpolateScale

.SH SYNOPSIS 
.B hev-interpolateScale  [ --s0 \fIs\fB ] [ --s1 \fIs\fB ]  [ --absolute ] [ --startTime \fIt\fB ] [ --duration \fIt\fB ] [ --frameRate \fIr\fB ] [ --deltaScale \fIs\fB ]

.SH OPTIONS

.IP "--s0 \fIs\fR"
specifies the initial scale. If not specified, 1.0 is used.

.IP "--s1 \fIs\fR"
specifies the final scale. If not specified, 1.0 is used.

.IP "--absolute"
specifies that the times used in the IRIS control commands be absolute
times. i.e. OSG reference times. The default is to use time values that are
relative to the time the control command was processed.

.IP "--startTime \fIt\fR"
specifies the initial time offset in seconds to be added to all IRIS control
commands.  The default value is 0.

.IP "--duration \fIt\fR"
specifies the number of seconds to scale from s0 to s1. If not supplied,
duration will be set to 1.0 second.

.IP "--frameRate \fIr\fR"
specifies the number of times per second \fBhev-interpolateScale\fR
will interpolate between the scales. If \fIframeRate\fR is not specified, 60 will be used.

.IP "--deltaScale \fIs\fR"
specifies the amount that the navigation node's scale value needs to
change before a command is written. The amount is the geometric difference from the last
written scale, or its inverse, whichever is largest.

If \fIdeltaScale\fR is not specified, it defaults to .001.

If \fIdeltaScale\fR is set to zero a command containing a scale
value will be written \fIframeRate\fR times a second.

.SH DESCRIPTION

\fBhev-interpolateScale\fR will write IRIS control commands that will
modify the navigation node's scale to go from \fIs0\fR to
\fIs1\fR over the course of \fIduration\fR seconds, writing up to \fIframerate\fR commands
per second.

\fBhev-interpolateScale\fR liberally comments the control commands it
generates. Like any IRIS control command file, the results can be edited with any text
editor.

.SH EXAMPLE

hev-interpolateScale --s0 1 --s1 3  --start 10 --duration 2.5  --frameRate 4 > /tmp/scoot.iris

will generate and store in the file /tmp/scoot.iris:

AFTER TIME 10.000000 NAV SCALE 1.000000
.br
AFTER TIME 10.250000 NAV SCALE 1.200000
.br
AFTER TIME 10.500000 NAV SCALE 1.400000
.br
AFTER TIME 10.750000 NAV SCALE 1.600000
.br
AFTER TIME 11.000000 NAV SCALE 1.800000
.br
AFTER TIME 11.250000 NAV SCALE 2.000000
.br
AFTER TIME 11.500000 NAV SCALE 2.200000
.br
AFTER TIME 11.750000 NAV SCALE 2.400000
.br
AFTER TIME 12.000000 NAV SCALE 2.600000
.br
AFTER TIME 12.250000 NAV SCALE 2.800000
.br
AFTER TIME 12.500000 NAV SCALE 3.000000



.SH SEE ALSO

hev-timeline, hev-navigationToInterpolate, hev-timelineToIris, hev-interpolateRotation, hev-interpolateTranslation, hev-readNav

.PP
John Kelso, kelso@nist.gov
.br
NIST High Performance Computing and Visualization Group
.br
November 2011
.br
\fBhttp://www.nist.gov/mcsd/hpcvg/\fR



idea/src/hev-timeline/hev-interpolateScale.cpp
#include <stdio.h>
#include <iostream>
#include <stdlib.h>

#include <iris.h>

#include "utils.h"

void usage()
{
    fprintf(stderr,"Usage: hev-interpolateScale [ --s0 s ] [ --s1 s ] [ --absolute ] [ --startTime t ] [ --duration ] [ --frameRate ] [ --deltaScale s ]\n") ;
}

////////////////////////////////////////////////////////////////////////
int main(int argc, char** argv)
{

    // send dtk messages to stdout
    dtkMsg.setFile(stderr) ;

    double s0 = 1 ;
    double s1 = 1 ;
    float startTime = 0 ;
    float duration = 1 ;
    float frameRate = -1 ;
    bool absolute = false ;
    double deltaScale = .001 ;

    int c = 1 ;
    while (c < argc && argv[c][0] == '-' && argv[c][1] == '-')
    {
	if (c>=argc || iris::IsSubstring("--absolute",argv[c],3))
	{
	    c++ ;
	    absolute = true ;
	}
	else if (c>=argc || iris::IsSubstring("--startTime",argv[c],4))
	{
	    c++ ;
	    if (c>=argc || !iris::StringToFloat(argv[c],&startTime) || startTime<0) 
	    {
		dtkMsg.add(DTKMSG_ERROR, "hev-interpolateScale: invalid startTime value\n") ;
		usage() ;
		return 1 ; 
	    }
	    c++ ;
	}
	else if (c>=argc || iris::IsSubstring("--duration",argv[c],4))
	{
	    c++ ;
	    if (c>=argc || !iris::StringToFloat(argv[c],&duration) || duration<=0) 
	    {
		dtkMsg.add(DTKMSG_ERROR, "hev-interpolateScale: invalid duration value\n") ;
		usage() ;
		return 1 ; 
	    }
	    c++ ;
	}
	else if (c>=argc || iris::IsSubstring("--frameRate",argv[c],4))
	{
	    c++ ;
	    if (c>=argc || !iris::StringToFloat(argv[c],&frameRate) || frameRate<=0) 
	    {
		dtkMsg.add(DTKMSG_ERROR, "hev-interpolateScale: invalid frameRate value\n") ;
		usage() ;
		return 1 ; 
	    }
	    c++ ;
	}
	else if (c>=argc || iris::IsSubstring("--deltaScale",argv[c],3))
	{
	    c++ ;
	    if (c>=argc || !iris::StringToDouble(argv[c],&deltaScale) || deltaScale <0) 
	    {
		dtkMsg.add(DTKMSG_ERROR, "hev-interpolateScale: invalid deltaScale value\n") ;
		usage() ;
		return 1 ; 
	    }
	    c++ ;
	}
	else if (c>=argc || !strcmp("--s0",argv[c]))
	{
	    c++ ;
	    if (c>=argc || !iris::StringToDouble(argv[c],&s0)) 
	    {
		dtkMsg.add(DTKMSG_ERROR, "hev-interpolateScale: invalid --s0 value\n") ;
		usage() ;
		return 1 ; 
	    }
	    c++ ;
	}
	else if (c>=argc || !strcmp("--s1",argv[c]))
	{
	    c++ ;
	    if (c>=argc || !iris::StringToDouble(argv[c],&s1)) 
	    {
		dtkMsg.add(DTKMSG_ERROR, "hev-hev-interpolateScale: invalid --s1 value\n") ;
		usage() ;
		return 1 ; 
	    }
	    c++ ;
	}
	else 
	{
	    dtkMsg.add(DTKMSG_ERROR, "hev-hev-interpolateScale: unrecognized option: %s\n",argv[c]) ;
	    usage() ;
	    return 1 ;
	}
    }

#if 0
    // get the frame rate from the system if frameRate not passed as a parameter
    if (frameRate < 0) frameRate = getDefaultFrameRate() ;
#else
    if (frameRate < 0) frameRate = 60 ;
#endif

    std::string timeString ;
    if (absolute) 
    {
	timeString = "ABSTIME" ;
	printf("# time is relative to start of first graphics frame\n") ;
    }
    else 
    {
	timeString = "TIME" ;
	printf("# time is relative to when control commands are received\n") ;
    }
    
    printf("# scaling from s0 = %f to s1 = %f; start time = %f, duration = %f, frame rate = %f\n",s0, s1, startTime, duration, frameRate) ;

    Posn p0, p1;
    p0.scale = s0 ;
    p0.t = startTime ;
    p1.scale = s1 ;
    p1.t = startTime + duration ;

    Interpolate interp(p0, p1) ;

    // always write the first
    printf("# first t = %f\n",p0.t) ;
    printf("AFTER %s %f NAV SCALE %f\n",timeString.c_str(), p0.t, p0.scale) ;
    Posn lastPosn = p0 ;

    for (int i=1; i<frameRate*duration; i++)
    {
	double t = p0.t + double(i)/frameRate ;
	Posn interpPosn ;
	printf("# interpolated t = %f\n",t) ;
	if (!interp.at(t, &interpPosn))
	{
	    dtkMsg.add(DTKMSG_ERROR, "hev-interpolateScale: interpolation error at t=%f\n",t) ;
	}
	else
	{
	    // only write if it changed from last time
	    if (movedEnoughScale(interpPosn.scale, lastPosn.scale, deltaScale))
	    {
		lastPosn.scale = interpPosn.scale ;
		printf("AFTER %s %f NAV SCALE %f\n",timeString.c_str(), t, interpPosn.scale) ;
		fflush(stdout) ;
	    }
	}
    }
    // always write the last
    printf("# last t = %f\n",p1.t) ;
    printf("AFTER %s %f NAV SCALE %f\n",timeString.c_str(), p1.t, p1.scale) ;
    return 0 ;
}
